[prometheusremotewriteexporter] reduce allocations in createAttributes

edma2 opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

While profiling the collector, we found that the `createAttributes` function was responsible for a significant chunk of allocations (30-40%) which was leading to a high CPU usage spent in GC. 

`createAttributes` is responsible for converting attributes of a given data point to Prometheus labels. For simplicity, it allocates a new labels slice for every data point. We found that reducing allocations here significantly reduced GC time in our environment (in some deployments as much as ~50%).

The strategy in this PR is to reuse the slice array as much as possible. The backing array will automatically resize as needed (batching with a batch processor will effectively set an upper bound). Note: we don't need to synchronize access to this (e.g. `sync.Pool`) since [the exporter is configured with 1 consumer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c52a9a7ae1563127254530d416a8c569c54750bd/exporter/prometheusremotewriteexporter/factory.go#L69).


**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

Modified unit tests and ran benchmarks locally.
Works in our production environment.

benchstat output
```
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
                    │ /tmp/old.txt  │            /tmp/new.txt             │
                    │    sec/op     │   sec/op     vs base                │
CreateAttributes-16   1010.0n ± 25%   804.0n ± 7%  -20.40% (p=0.000 n=10)

                    │ /tmp/old.txt │            /tmp/new.txt            │
                    │     B/op     │    B/op     vs base                │
CreateAttributes-16    371.00 ± 0%   91.00 ± 0%  -75.47% (p=0.000 n=10)

                    │ /tmp/old.txt │            /tmp/new.txt            │
                    │  allocs/op   │ allocs/op   vs base                │
CreateAttributes-16     7.000 ± 0%   5.000 ± 0%  -28.57% (p=0.000 n=10)
```

**Documentation:** <Describe the documentation added.>

dashpole commented:
cc @jmichalek132 @ArthurSens

dashpole commented on pkg/translator/prometheusremotewrite/metrics_to_prw.go:
```diff
@@ -27,8 +27,7 @@ type Settings struct {
 // FromMetrics converts pmetric.Metrics to Prometheus remote write format.
-func FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
-	c := newPrometheusConverter()
+func (c *PrometheusConverter) FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
```
we consider this a breaking API change in for changelog purposes.

edma2 replied:
updated changelog

dashpole commented on pkg/translator/prometheusremotewrite/metrics_to_prw.go:
```diff
@@ -39,21 +38,37 @@ func FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.Time
+// best to keep it around for the lifetime of the Go process. Due to this shared
+// state, PrometheusConverter is NOT thread-safe and is only intended to be used by
+// a single go-routine at a time.
+// Each FromMetrics call should be followed by a Reset when the metrics can be safely
```
Should we emit a warning log or something if someone calls FromMetrics without Resetting?

edma2 replied:
Moved the reset to always be called inside FromMetrics so this is no longer a user concern.

ArthurSens replied:
Now that the user doesn't need to call reset, should we remove this part of the comment?
`// Each FromMetrics call should be followed by a Reset.....`

dashpole commented on pkg/translator/prometheusremotewrite/metrics_to_prw.go:
```diff
@@ -27,8 +27,7 @@ type Settings struct {
 // FromMetrics converts pmetric.Metrics to Prometheus remote write format.
-func FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
-	c := newPrometheusConverter()
+func (c *PrometheusConverter) FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
```
Now that users create a struct before calling FromMetrics, we should consider moving any configuration or settings to the creation of the converter to simplify usage.

edma2 replied:
This makes sense, but now it's in global sync.Pool rather than tied to a specific exporter instance so it needs to be reusable across instances.

ArthurSens commented:
> Note: we don't need to synchronize access to this (e.g. `sync.Pool`) since [the exporter is configured with 1 consumer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c52a9a7ae1563127254530d416a8c569c54750bd/exporter/prometheusremotewriteexporter/factory.go#L69).

We don't plan to keep this forever, right? Ideally we'll be able to shard this to improve throughput, we're just hardcoding this to 1 because OTel's exporter helper doesn't ensure ordering.

On the other hand, I agree that we shouldn't block optimizations based on something we want to do in the future 😬. @edma2, knowing that we'll eventually shard the output, any suggestions on how to do this without sacrificing your optimization?

jmichalek132 commented:
> > Note: we don't need to synchronize access to this (e.g. `sync.Pool`) since [the exporter is configured with 1 consumer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c52a9a7ae1563127254530d416a8c569c54750bd/exporter/prometheusremotewriteexporter/factory.go#L69).
> 
> We don't plan to keep this forever, right? Ideally we'll be able to shard this to improve throughput, we're just hardcoding this to 1 because OTel's exporter helper doesn't ensure ordering.
> 
> On the other hand, I agree that we shouldn't block optimizations based on something we want to do in the future 😬. @edma2, knowing that we'll eventually shard the output, any suggestions on how to do this without sacrificing your optimization?

I wonder also since you can have multiple pipelines with multiple remote write exporters (i.e. sending data from dev cluster to 2 destination, dev and prod) if that would break this too.

edma2 commented:
> On the other hand, I agree that we shouldn't block optimizations based on something we want to do in the future 😬. @edma2, knowing that we'll eventually shard the output, any suggestions on how to do this without sacrificing your optimization?

@ArthurSens my initial thought here is maybe wrap things in a `sync.Pool` but I'm not sure how that might affect performance. I'm kinda leaning to making that a future change but I can think about it more.

> I wonder also since you can have multiple pipelines with multiple remote write exporters (i.e. sending data from dev cluster to 2 destination, dev and prod) if that would break this too.

Each exporter would have its own instance of `PrometheusConverter` so I think it would be ok.

edma2 commented:
@ArthurSens @dashpole @jmichalek132 I addressed comments and also changed the implementation so it's now in a sync.Pool. This now supports concurrent access from the exporter class in case it ever supports more than 1 worker at a time.

edma2 committed:
[prometheusremotewriteexporter] reduce allocations in createAttributes (#57)

createAttributes was allocating a new label slice for every series, which generates mucho garbage (~30-40% of all allocations). Keep around a re-usable underlying array of labels to reduce allocations on the hot path.

edma2 committed:
add changelog

edma2 committed:
wrap converters in sync.Pool

edma2 committed:
change type: breaking

edma2 committed:
move Reset() to reset internal call

edma2 committed:
fix v2 remote write

ArthurSens commented:
Awesome edma! I'm struggling a bit to find time to review this one, just wanted to let you know that this is on my list :)

ArthurSens commented on pkg/translator/prometheusremotewrite/helper.go:
```diff
@@ -96,46 +97,40 @@ var seps = []byte{'\xff'}
-
 	// map ensures no duplicate label name
-	l := make(map[string]string, maxLabelCount)
+	l := c.labelsMap
+	clear(l)
```
Could we also clear `labelMap` when we call `reset()`? I think it would be cleaner if we reset the state in one single place instead, or is there any particular reason to do it here?

ArthurSens replied:
That makes sense, thank you! Could we add a comment explaining it?

ArthurSens commented on pkg/translator/prometheusremotewrite/helper.go:
```diff
@@ -175,12 +170,12 @@ func createAttributes(resource pcommon.Resource, attributes pcommon.Map, externa
 	}
 
-	return labels
+	return c.labels[startIndex:]
```
Is there any possibility that `len(c.labels)` is not 0 here? It's reset every time we call FromMetrics and I couldn't find any other place in the code where we write to this array, so why not just return `c.labels` and not worry about startIndex? I might be missing something but it feels like we're overcomplicating things here

```go
for k, v := range l {
	c.labels = append(c.labels, prompb.Label{Name: k, Value: v})
}
return c.labels
```

edma2 replied:
`startIndex` is important for keeping the returned slices isolated from each other while sharing the same underlying array within a single `FromMetrics` call. It is 0 only for the first series of a batch.

Here is how it works: `FromMetrics` is called once per batch, and `createAttributes` for every series within the batch. We want to re-use the backing array of the `labels` slice for all series within a single batch. We do that by appending the labels of each series to the end of the slice. Finally we return only starting from `startIndex` so the caller doesn't see labels from other series (while reusing the same backing array which naturally grows up to the size needed to fit a single `FromMetrics` call).

For example, if X1...X4 are labels from series X and Y1...Y3 are labels from series Y, then the backing array of `c.labels` will look like `[X1, X2, X3, X4, Y1, Y2, Y3]`. Then on the next `FromMetrics` call the index is reset to 0 and we can re-use the entire array with zero allocations.

ArthurSens replied:
Perfect, thank you for the explanation :)

Now what I'm thinking is if we have tests that assure concurrency works. Mostly to make sure we don't break the thread-safe promise by accident

edma2 committed:
update  comments

ArthurSens commented on .chloggen/prometheusremotewrite-optimize-createattributes.yaml:
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'breaking'
```
That's an enhancement, right? Not a breaking change

ArthurSens reviewed (commented):
Alright, thanks for the patience and for kindly explaining everything here :) 

Regarding all the code so far, it LGTM! Let's just add an extra test that certifies we're not calling FromMetrics concurrently in other areas of the codebase. This test will probably live in the exporters that uses the translator pkg, e.g. prometheusremotewriteexporter

Maybe something like this(pseudo-code):

```go
func generateMetric(i int) pmetric.Metrics {
    m := GenerateMetricsOneMetric()
    m.ResourceMetrics().At(0).ScopeMetrics().At(0).Metrics().At(0).Sum().DataPoints().AppendEmpty().SetIntValue(i)
    return m
}

func TestPushMetricsConcurrency(t testint.T) {
  // Create mock where prwexporter sends metric to
  // Create prwexporter
  n := 1000
  metrics := make([]pmetric.Metrics, 0, n)
   for i<n {
    metrics = append(metrics, generateMetric(i)
  } 

  var wg sync.WaitGroup
  wg.Add(n)
  for range metrics {
      go func() {
          prwexporter.PushMetrics(background, metrics[i])
          wg.Done()
      }()
  }
  wg.Wait()

  //assert Metrics arrived in the mock.
}
```

edma2 committed:
Merge branch 'main' into prometheusremotewrite-optimize-createattributes

edma2 committed:
fix bad merge

edma2 committed:
add concurrency test

edma2 commented:
@ArthurSens thank you for the suggestion. Added a test, PTAL

edma2 committed:
fix lint errors

edma2 committed:
move test to a separate file due to unrelated race condition

edma2 commented on exporter/prometheusremotewriteexporter/exporter_concurrency_test.go:
```diff
@@ -0,0 +1,153 @@
+//go:build !race
+// +build !race
+
+// note: this test doesn't pass currently due to a race condition in batchTimeSeries
```
@ArthurSens I discovered a race condition when multiple goroutines run which seems unrelated to my change. The batching logic uses shared state which won't work with multiple goroutines. I added some build tags above to exclude this test until it's fixed.

ArthurSens replied:
Hmmm, I've run your test on the main branch, and indeed, we have a data race like you said :O 

I am not sure how to proceed here. To be honest, I think we need to fix the test first, so we're safe to merge this one, too

ArthurSens replied:
I've opened a PR trying to address the issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36524

Wanna share your opinion? :)

edma2 committed:
lint: move header

ArthurSens commented:
This is on my review list; I'll get to it hopefully tomorrow or early next week 🤞

ArthurSens commented on pkg/translator/prometheusremotewrite/metrics_to_prw.go:
```diff
@@ -27,8 +27,11 @@ type Settings struct {
+// Only call this function when the result of the previous call is no longer needed
+// (i.e. has been exported out of process). The return values of this function depend on
+// internal state which is reset on every call.
+func (c *PrometheusConverter) FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
```
Although I like the idea of reimagining the API, we have way too much activity happening in this component at the moment, and it is mostly coming from new members like myself, @edma2, and @jmichalek132. 

The OTel-Prometheus SIG group is relatively small, and I'd love to optimize contributions for community growth. Once we have enough hands, we can plan the refactoring you are proposing. By optimizing contributions for community growth, I mean fast interactions, try our best to keep PRs open for max 2 weeks (to avoid rebases), accepting imperfect PRs while encouraging improvements in follow-ups, etc.

What do you think?

Aneurysm9 replied:
I never said anything about reimagining an API.  I said that if we're going to expose the `PrometheusConverter` type we need to think about what that exposed API should be and I submit that simply taking this helper method and making it the public API ain't it.  The API is already there, simply not exported.  If a `Reset()` method needs to be added, that's fine.  What we shouldn't do is export this helper method, which has the wrong signature and does more work than is needed in order to keep a separation between a performance optimization and the public API of this package, simply because it's the "efficient" thing to do from the view of how long PRs remain open.  That will mean a breaking change now and another breaking change when a sensible API is actually exposed.

An alternative that wouldn't involve considering the shape of a public API at all, and wouldn't involve any breaking changes to any public API, would be to move the `converterPool` out of the exporter and into the translation package.  The existing `FromMetrics` function is the only thing that needs it and it only needs to obtain a converter, use it, and return it to the pool.  As this PR currently stands the exporter never uses a `PrometheusConverter` instance for anything more than a single invocation of `FromMetrics`.  Doing it this way should make this PR considerably smaller and eliminates any public API changes.

What do you think?

edma2 replied:
> An alternative that wouldn't involve considering the shape of a public API at all, and wouldn't involve any breaking changes to any public API, would be to move the converterPool out of the exporter and into the translation package. The existing FromMetrics function is the only thing that needs it and it only needs to obtain a converter, use it, and return it to the pool. As this PR currently stands the exporter never uses a PrometheusConverter instance for anything more than a single invocation of FromMetrics. Doing it this way should make this PR considerably smaller and eliminates any public API changes.

@Aneurysm9 This wouldn't work because the translator package (i.e. FromMetrics) doesn't know when to put the converter back in the pool. It can't do it at the end of FromMetrics because the metrics may not have been shipped off yet; putting it back in the pool may prematurely reset its state and send invalid attributes.

Aneurysm9 replied:
Then we're back to what I said initially: expose the correct API for the `PrometheusConverter` type, don't simply shove this shim into it.  If we're going to export that type it should have the following interface:

```go
type PrometheusConverter interface {
  FromMetrics(pmetric.Metrics, Settings) error
  TimeSeries() []prompb.TimeSeries
  Reset()
}
```

It doesn't necessarily need to be an interface, and `TimeSeries()` returning a `prompb` type complicates the use case where an interface would actually make sense, but that's the shape of the exported methods I would expect.  They're all there.  Let's export them rather than perpetuating the stopgap that's in place and making it take multiple breaking changes to get there.

edma2 replied:
@Aneurysm9 @ArthurSens apologies for the long delay, I was having trouble finding time to carefully review the feedback. This solution has been working well for us in production but I'd love to upstream it!

After looking at the code and prior PRs, I think I understand better what you were getting at. I made the following changes
* Cleaned up PrometheusConverter API: it now exposes `Convert()`, `TimeSeries()`, and `Reset()` functions, similar to your suggestion.
* Got rid of `FromMetrics`: all it was doing was restructuring the slice from `TimeSeries()` as a map with indices as keys. The only call site in this repository, the PRW exporter, doesn't care about the keys so it just needs to call `TimeSeries()`. I can also keep `FromMetrics` as a shim to preserve backwards compatibility if needed (internally it would use PrometheusConverter but not be exposed as a method receiver).

What do you think?

edma2 commented:
@ArthurSens can we re-open this PR? I don't think I have the permissions.

edma2 committed:
no-op

atoulme commented:
Please resolve conflicts. Moving to draft while this is being worked on.

edma2 committed:
Simplify PrometheusConverter interface

FromMetrics no longer needed.

edma2 committed:
Merge branch 'main' into prometheusremotewrite-optimize-createattributes

edma2 committed:
update wal_test.go

edma2 commented:
@atoulme I've updated my branch and merged with `main` but I'm not seeing my changes reflected here. Can you please re-open this PR?

edma2 committed:
Merge branch 'main' into prometheusremotewrite-optimize-createattributes

ArthurSens commented:
Hi @edma2 , thanks for continuously coming back to this work! I've been struggling to find time to review this PR again, I just wanted to let you know that I'm aware of its existence and that I'm trying to allocate time for the review!

atoulme commented:
I see one more conflict on the code - please resolve and mark ready for review again.

edma2 committed:
Merge branch 'main' into prometheusremotewrite-optimize-createattributes

edma2 committed:
add back newline

ArthurSens commented:
@edma2, to be fully transparent and respectful of your time, let's put the work here on hold for a while. The optimization you're doing is excellent, but it's also not easy to understand without paying a decent amount of attention.

The team's priority right now is adhering to the OTel->Prometheus specification and implementing version 2 of the Remote Write Protocol, which is creating many conflicts in your PR.

I'm a bit concerned that what I'm saying will put you in a bad mood, and I apologize for that, but I think it would be worse if I don't say anything and let you keep resolving Merge conflicts every single week 😕

edma2 commented:
@ArthurSens no problem, I totally understand if there are higher priorities right now. Thanks for giving me a heads up. Do you know when would be a good time to revisit the PR? Also, if splitting the PR into smaller pieces would make reviewing it easier, I can do that.

ArthurSens commented:
I'd say that after we solve https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33661, it should be a good time to get back to this. We shouldn't see multiple merge conflicts after that
